// infobip/lib/transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("request");
class Transport {
    constructor(config) {
        this.config = config;
    }
    send(options, callback) {
        let headers = options.headers || {};
        let method = options.method || "post";
        let path = options.path;
        let url = this.config.baseUrl + path;
        let body = options.body;
        // Default headers
        headers["Authorization"] = this.config.getAuthorization();
        headers["Content-Type"] = "application/json";
        headers["Accept"] = "application/json";
        let reqOptions = {
            url: url,
            json: true,
            body: body,
            headers: headers
        };
        request[method](reqOptions, (err, response, body) => {
            this.handleResponse(err, response, body, callback);
        });
    }
    handleResponse(err, res, resBody, callback) {
        if (typeof callback !== "function") {
            return;
        }
        let body = resBody ? resBody : null;
        callback(err, res, body);
    }
}
exports.default = Transport;
